### PR TITLE
Set pod affinity for host assisted clone

### LINKED ISSUE #2576: Set pod affinity for host assisted clone source pods
**Is your feature request related to a problem? Please describe**:
Basically, the issue in https://github.com/kubernetes/kubernetes/issues/103305 applies to CDI
in the concurrent host assisted clone use case, on storage that is not topology constrained (globally accessible from all nodes).
Consider the scenario:
- Initiate 2 concurrent host assisted clone from source-pvc that was provisioned by rook-ceph-block with `ReadWriteOnce`
- It is entirely possible that each of the pods lands on a different node, and thus one will remain pending

**Describe the solution you'd like**:
The suggested solution in https://github.com/hetznercloud/csi-driver/issues/319#issuecomment-1324052836

**Describe alternatives you've considered**:

**Additional context**:
This is not a problem for topology constrained storage like HPP because scheduler will ensure
the pods land where the PV is (`volume.kubernetes.io/selected-node` ann)

### PR DESCRIPTION
**Pull request by ido106**
Signed-off-by: Ido Aharon <iaharon@redhat.com>

**What this PR does / why we need it**:
Pods using the same PVC are scheduled on different nodes. We expect them to be scheduled on the same K8S node but they don't and thus cannot be attached to the PersistentVolume. The solution is to set [pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2576

**Special notes for your reviewer**:

**Release note**:

```release-note
Set pod affinity for host assisted clone source pods, in order to schedule them on the same K8S node.
```

### CODE REVIEW

File: tests/cloner_test.go [hunk 1]
```diff
@@ -2911,6 +2915,8 @@ func doFileBasedCloneTest(f *framework.Framework, srcPVCDef *v1.PersistentVolume
 	targetDV.Labels["test-label-1"] = "test-label-key-1"
 	targetDV.Annotations["test-annotation-1"] = "test-annotation-key-1"
 
+	targetDV.Annotations[controller.AnnPodRetainAfterCompletion] = "true"
```
Comment (akalenyu) [inline tests/cloner_test.go:2918]: This already happens in the test at https://github.com/kubevirt/containerized-data-importer/blob/597076f7d7c322579b45c191ae21c751f357a96f/tests/cloner_test.go#L119 Do we still want to keep?

**Review by awels — COMMENTED** [review]
Can we also verify in a functional test that when we clone multiple times from the same source, and the source is RWO, that all the source pods are on the same node? We already have a multiple clone test [here](https://github.com/kubevirt/containerized-data-importer/blob/main/tests/cloner_test.go#L1496-L1539)

We just need to modify it to check if all the pods are on the same node IF the source volume is RWO (which is not always the case, nfs and ceph will be RWX)

File: tests/cloner_test.go (hunk 1 above)
Comment (ido106) [reply tests/cloner_test.go:2918]: I think it is enough. I will delete it and run the tests again.

File: pkg/controller/clone-controller.go [hunk 2]
```diff
@@ -532,10 +534,13 @@ func (r *CloneReconciler) CreateCloneSourcePod(image, pullPolicy string, pvc *co
...
 
 // MakeCloneSourcePodSpec creates and returns the clone source pod spec based on the target pvc.
-func MakeCloneSourcePodSpec(sourceVolumeMode corev1.PersistentVolumeMode, image, pullPolicy string, imagePullSecrets []corev1.LocalObjectReference, sourcePvcName, sourcePvcNamespace, ownerRefAnno string,
-	serverCACert []byte, targetPvc *corev1.PersistentVolumeClaim, resourceRequirements *corev1.ResourceRequirements,
+func MakeCloneSourcePodSpec(sourceVolumeMode corev1.PersistentVolumeMode, image, pullPolicy string, imagePullSecrets []corev1.LocalObjectReference, ownerRefAnno string,
+	serverCACert []byte, targetPvc, sourcePvc *corev1.PersistentVolumeClaim, resourceRequirements *corev1.ResourceRequirements,
 	workloadNodePlacement *sdkapi.NodePlacement) *corev1.Pod {
 
+	sourcePvcName := sourcePvc.GetName()
+	sourcePvcNamespace := sourcePvc.GetNamespace()
```
[the remark this answers is not in the document]
Comment (ido106) [reply pkg/controller/clone-controller.go:542]: `sourcePvcName` and `sourcePvcNamespace` were originally sent as parameters to the function, so I decided to pass `sourcePvc` instead and place them at the beginning of the function for backwards compatibility. That was after I used `sourcePvcUid`, but I thought about it afterwards and I really should define it in the same way like you said.

**Comment by akalenyu** [comment]
> Can we also verify in a functional test that when we clone multiple times from the same source, and the source is RWO, that all the source pods are on the same node? We already have a multiple clone test [here](https://github.com/kubevirt/containerized-data-importer/blob/main/tests/cloner_test.go#L1496-L1539)
> 
> We just need to modify it to check if all the pods are on the same node IF the source volume is RWO (which is not always the case, nfs and ceph will be RWX)

So this test actually doesn't test ceph with host assisted (only smart clone), that is why I made https://github.com/kubevirt/containerized-data-importer/pull/2575
(https://github.com/kubevirt/containerized-data-importer/issues/2576 is only interesting in case the storage is not topology constrained).

I canceled the hold on that PR so we can decide if we want it in

**Comment by ido106** [comment]
> Changes look good, but I think you might have missed my overall comment about verifying that when we do multiple clones at once (we have tests that already do this) we should ensure that the source pods are all on the same node. I put in a link to the tests.

The problem is that this test will work anyway even without these changes. Maybe I can create 2 clones with 2 different source pvc's and check if each one was placed on the correct node ?
I added the test you asked for (locally) so I can push it but maybe we should check it another way.

File: tests/cloner_test.go [hunk 3]
```diff
@@ -1813,45 +1823,57 @@ var _ = Describe("all clone tests", func() {
...
 				Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(dv), matchFile, md5sum[:32])).To(BeTrue())
+
+				pvc, err := f.K8sClient.CoreV1().PersistentVolumeClaims(dv.Namespace).Get(context.TODO(), dv.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				if cloneSourcePod := pvc.Annotations[ctrl.AnnCloneSourcePod]; cloneSourcePod != "" {
+					By(fmt.Sprintf("Getting pod %s/%s", dv.Namespace, cloneSourcePod))
+					pod, err := f.K8sClient.CoreV1().Pods(dv.Namespace).Get(context.TODO(), cloneSourcePod, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					podsNodeName[dv.Name] = pod.Spec.NodeName
```
Comment (awels) [inline tests/cloner_test.go:1857]: Would it make sense to have the map be `map[string]bool` and on this line just say `podsNodeName[pod.Spec.NodeName]=true` and below check `Expect(podsNodeName).To(HaveLen(1))` We are essentially creating a set of node names, and ensure there is only 1 node name. This demonstrates all the pods were on the same node.

**Review by awels — COMMENTED** [review]
Looks good just one minor comment. was having a hard time understanding what that part of the test was doing.

File: tests/cloner_test.go [hunk 4]
```diff
@@ -1042,14 +1042,13 @@ var _ = Describe("all clone tests", func() {
...
+					// All pods should be in the same node except when the map is empty in smart clone
 					if len(podsNodeName) > 0 {
-						Expect(podsNodeName[targetDvs[0].Name]).To(Equal(podsNodeName[targetDvs[1].Name]))
-						Expect(podsNodeName[targetDvs[1].Name]).To(Equal(podsNodeName[targetDvs[2].Name]))
+						Expect(podsNodeName).To(HaveLen(1))
 					}
 				})
 
-				It("[rfe_id:1277][test_id:1899][crit:High][vendor:cnv-qe@redhat.com][level:component] Should allow multiple cloning operations in parallel for block devices", func() {
+				FIt("[rfe_id:1277][test_id:1899][crit:High][vendor:cnv-qe@redhat.com][level:component] Should allow multiple cloning operations in parallel for block devices", func() {
```
Comment (akalenyu) [inline tests/cloner_test.go:1051]: Other than the FIts the PR looks good to me!

Comment (ido106) [reply tests/cloner_test.go:1051]: whoops

File: tests/cloner_test.go [hunk 5]
```diff
@@ -1019,19 +1020,37 @@ var _ = Describe("all clone tests", func() {
...
+						}
+
 						By("Deleting verifier pod")
 						err = f.K8sClient.CoreV1().Pods(f.Namespace.Name).Delete(context.TODO(), utils.VerifierPodName, metav1.DeleteOptions{})
 						Expect(err).ToNot(HaveOccurred())
 						_, err = utils.WaitPodDeleted(f.K8sClient, utils.VerifierPodName, f.Namespace.Name, verifyPodDeletedTimeout)
 						Expect(err).ToNot(HaveOccurred())
 					}
+
+					// All pods should be in the same node except when the map is empty in smart clone
```
Comment (akalenyu) [inline tests/cloner_test.go:1045]: You should be able to use `cloneType` in this context to tell you if this is going to be smart clone or not
For example `cloneType != "snapshot"`

Comment (ido106) [reply tests/cloner_test.go:1045]: Ok, I changed the condition to `if cloneType == "network"`

**Comment by akalenyu** [comment]
/test pull-containerized-data-importer-e2e-ceph

Same issue as in https://github.com/kubevirt/containerized-data-importer/issues/2550,
maybe this change make it fail more often? in that case we might want to comment out `Expect(restartCount).To(BeNumerically("<", 2))` with a TODO to remove once the issue is fixed

**Comment by akalenyu** [comment]
forgot to lgtm again
/lgtm

**Pull request merged**